traits/trait_list_object.py

```diff
@@ -59,34 +59,117 @@ def __repr__(self):
 
 
 def _normalize_index(index, length):
-    """ Normalize integer index to range 0 to len (inclusive). """
+    """ Normalize integer index to range 0 to len (inclusive).
+
+    These rules reflect the normalization that Python uses for the
+    ``insert`` and ``pop`` operations.
+
+    """
+
     index = operator.index(index)
     if index < 0:
         return max(0, length + index)
     else:
         return min(length, index)
 
 
-def _normalize_slice(index, length):
-    """ Normalize slice start and stop to range 0 to len (inclusive). """
+def _normalize_slice_or_index(index, length):
+    """ Normalize a slice or index for use with __delitem__ or __setitem__.
 
-    # Non-extended slice (step = 1): return only need the start index.
-    if index.step is None or index.step == 1:
-        if index.start is None:
-            return 0
-        else:
-            return _normalize_index(index.start, length)
+    For slices with positive step, returns a slice that's equivalent for the
+    purposes of __delitem__ and __setitem__ operations. For slices with
+    negative step, a normalized slice representing the reverse of the given
+    slice is returned: note that in this case, the matching *added* and
+    *removed* lists will need to be reversed.
+
+    Slices with a step of 1 or -1 are normalized to a single integer index,
+    referring to the position of the first element referenced by the slice.
+
+    Similarly, slices that refer to only a single element of the corresponding
+    list (for example, a slice of `[1::10]` applied to a list of length 5)
+    are normalized to the index that refers to that same element.
+
+    Empty slices are also normalized to a single index. Note that in the case
+    of an empty slice, the corresponding __delitem__ or __setitem__ operation
+    does not cause any list change, so does not issue a notification. So the
```

Comment:
We may want to consider raising in this case: we shouldn't even end up _calling_ this function for an empty slice.

Reply:
That means we will have to capture the exception for this, just wondering what the performance impact this additional code path might bring and if that is worth it. It might be safer to still return an empty slice instead of an integer which will get you something from the list in the event of misuse?

Reply:
No, no need to capture. If the exception ever arose, that would be a coding bug.

Reply:
Actually, returning index is also fine persumably the `removed` and `added` would also be empty if the notification was not silenced, not so much misuse can happen there.

Reply:
And performance is highly unlikely to be a consideration here :-).  We should almost _never_ be making behaviour decisions based on the cost of an exception raise / capture / ...

Reply:
> Actually, returning index is also fine

Yes, but I think I'd rather raise as a defensive measure, since this case will only be exercised as a result of a bug elsewhere.

If we really _did_ want an index for this case, we'd need to think about normalization again: in the case of empty `removed` and `added`, an index of `0` is just as meaningful as an index of `2`. But since this isn't a case that should come up, I'd rather delay that decision.

Reply:
I guess I am not seeing why there is no need to capture if this function raises for empty slice. Maybe if you give this a go then I would understand what you mean.

Reply:
This function should never be called for an empty slice, because:

- an empty slice implies no list change has taken effect
- which implies that no notification will be issue
- which implies that we don't need a normalized index to pass to notifiers.

That does possibly mean shifting some code slightly so that the normalized index is only computed in the "I'm about to notify" branch.

Reply:
But yes, I'll update the code to show what I mean.

Reply:
> This function should never be called for an empty slice, because:
> 
> * an empty slice implies no list change will take effect: there will be no added or removed items
> * which implies that no notification will be issued
> * which implies that we don't need a normalized index to pass to notifiers.
> 
> That does possibly mean shifting some code slightly so that the normalized index is only computed in the "I'm about to notify" branch.

With https://github.com/enthought/traits/pull/1026 the "about to notify" branch is gone and notification is issued even if no actual change happens. From issue https://github.com/enthought/traits/issues/972 I understand that it is the desired behaviour

Reply:
Good point; I've commented on the #1026 issue to clarify.

Reply:
> * an empty slice implies no list change will take effect: there will be no added or removed items

This is wrong for `__setitem__`. `x[5:5] = [10, 11, 12]` _does_ have an effect. Sorry, all.